File: leetcode_python/flatten_binary_tree_to_linked_list.py
# Definition for a  binary tree node
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    # @param root, a tree node
    # @return nothing, do it in place
    def flatten(self, root):

        if not root:
            return

        def pre_order(node, previous):
            if previous:
                previous.right = node
                previous.left = None

            # recurse
            remember_right = node.right if node.right else None
            remember_left = node.left if node.left else None
            if remember_left:
                left = pre_order(remember_left, node)
            if remember_right:
                right = pre_order(remember_right, left if remember_left else node)

            # return the previous node for the next round
            if remember_right:
                return right
            elif remember_left:
                return left
            else:
                return node

        pre_order(root, None)
        return root

File: leetcode_cpp/flatten_binary_tree_to_linked_list.cpp
/**
 * Definition for binary tree
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:

    TreeNode* pre_order(TreeNode *node, TreeNode *previous) {

        // remember the daughters
        TreeNode* remember_left = (node -> left != NULL) ? node -> left : NULL;
        TreeNode* remember_right = (node -> right != NULL) ? node -> right : NULL;

        // rewire
        if (previous != NULL) {
            previous -> right = node;
            previous -> left = NULL;
        }

        TreeNode* left = NULL;
        TreeNode* right = NULL;

        // recurse
        if (remember_left)
            left = pre_order(remember_left, node);
        if (remember_right)
            right = pre_order(remember_right, (remember_left) ? left : node);

        // return previous for the next step of traversal
        if (remember_right) {
            return right;
        }
        else if (remember_left) {
            return left;
        }
        else {
            return node;
        }
    }

    void flatten(TreeNode *root) {

        if (root == NULL) {
            return;
        }

        pre_order(root, NULL);

    }
};
